src/GPE.jl

```diff
@@ -160,45 +145,38 @@ Initialise the marginal log-likelihood of Gaussian process `gp`.
 function initialise_mll!(gp::GPE{X,Y,M,K,P,D}) where {X,Y,M,K,P,D}
     μ = mean(gp.mean,gp.x)
     Σ = cov(gp.kernel, gp.x, gp.data)
-    gp.cK = P.name.wrapper(Σ + (exp(2*gp.logNoise) + 1e-5)*I)
+    gp.cK = init_cK(P, Σ + (exp(2*gp.logNoise) + 1e-5)*I)
     y = gp.y - μ
     gp.alpha = gp.cK \ y
     # Marginal log-likelihood
     gp.mll = -(dot(y, gp.alpha) + logdet(gp.cK) + log2π * gp.nobs) / 2
     gp
 end
+init_cK(::Type{<:PDMat}, m) = PDMat(m)
 
 """
     update_cK!(gp::GPE)
 
 Update the covariance matrix and its Cholesky decomposition of Gaussian process `gp`.
 """
-function update_cK!(gp::GPE{X,Y,M,K,P,D}) where {X,Y,M,K,P,D}
+function update_cK!(gp::GPE)
     old_cK = gp.cK
-    if P <: ElasticPDMat
-        Σbuffer = view(old_cK.mat)
-    else
-        Σbuffer = old_cK.mat
-    end
+    Σbuffer = mat(old_cK)
     cov!(Σbuffer, gp.kernel, gp.x, gp.data)
     noise = (exp(2*gp.logNoise) + 1e-5)
     for i in 1:gp.nobs
         Σbuffer[i,i] += noise
     end
-    if P <: ElasticPDMat
-        chol_buffer = view(old_cK.chol).factors
-    else
-        chol_buffer = old_cK.chol.factors
-    end
+    chol_buffer = cholfactors(old_cK)
     copyto!(chol_buffer, Σbuffer)
     chol = cholesky!(Symmetric(chol_buffer))
-    if !(P <: ElasticPDMat)
-        gp.cK = P.name.wrapper(Σbuffer, chol)
-    end
+    gp.cK = wrap_cK(gp.cK, Σbuffer, chol)
     gp.cK
 end
 
-is_data_updated(gp::GPE{X,Y,M,K,P,D}) where {X,Y,M,K,P <: ElasticPDMat,D} = false
+wrap_cK(cK::PDMat, Σbuffer, chol) = PDMat(Σbuffer, chol)
+mat(cK::PDMat) = cK.mat
+cholfactors(cK::PDMat) = cK.chol.factors
```

Comment:
Again thinking ahead to other possible kinds of PDMats (like sparse GPs or using decompositions other than Cholesky), it's not clear that only these three steps will be different. So in this case we *should* copy-paste the `update_cK!` function and modify just the lines that are different.

Reply:
I am not totally happy with copy-pasting the full `update_cK!` function and dispatching on the level of `update_cK!`. For example, also here I guess the `1e-5` will be refactored at some point...
What about keeping `update_cK` as it is now and refactor it, when we know exactly what the needs are for e.g. sparse GPs? If everything is different there, one can of course dispatch on the level of `update_cK` to differentiate sparse and dense GPs and inside dense `update_cK` dispatch to differentiate elastic and non-elastic GPs, as it is now. What do you think?

Reply:
That seems alright, you're right that we can revisit this later.